server/src/main/java/org/elasticsearch/index/query/InnerHitContextBuilder.java

```diff
@@ -114,6 +115,16 @@ protected void setupInnerHitsContext(
         if (innerHitBuilder.getSorts() != null) {
             Optional<SortAndFormats> optionalSort = SortBuilder.buildSort(innerHitBuilder.getSorts(), searchExecutionContext);
             if (optionalSort.isPresent()) {
+                if (innerHitBuilder.getRescoreBuilders() != null) {
```

Comment:
This should be validated in the configuration object up in InnerHitBuilder. Otherwise, we will likely do a ton of work only to throw later when the context is actually built.